File: Source/AlphaTab/Rendering/Layout/PageViewLayout.cs
using System;
using AlphaTab.Collections;
using AlphaTab.Platform;
using AlphaTab.Platform.Model;
using AlphaTab.Rendering.Staves;
using AlphaTab.Util;

namespace AlphaTab.Rendering.Layout
{
    /// <summary>
    /// This layout arranges the bars into a fixed width and dynamic height region.
    /// </summary>
    internal class PageViewLayout : ScoreLayout
    {
        // left top right bottom
        public static readonly float[] PagePadding =
        {
            40, 40, 40, 40
        };

        public const float GroupSpacing = 20;

        private FastList<StaveGroup> _groups;
        private FastList<MasterBarsRenderers> _allMasterBarRenderers;
        private FastList<MasterBarsRenderers> _barsFromPreviousGroup;
        private float[] _pagePadding;

        public override string Name => "PageView";

        public PageViewLayout(ScoreRenderer renderer)
            : base(renderer)
        {
            _barsFromPreviousGroup = new FastList<MasterBarsRenderers>();
        }

        protected override void DoLayoutAndRender()
        {
            _pagePadding = Renderer.Settings.Display.Padding;
            if (_pagePadding == null)
            {
                _pagePadding = PagePadding;
            }

            if (_pagePadding.Length == 1)
            {
                _pagePadding = new[]
                {
                    _pagePadding[0], _pagePadding[0], _pagePadding[0], _pagePadding[0]
                };
            }
            else if (_pagePadding.Length == 2)
            {
                _pagePadding = new[]
                {
                    _pagePadding[0], _pagePadding[1], _pagePadding[0], _pagePadding[1]
                };
            }

            var x = _pagePadding[0];
            var y = _pagePadding[1];
            Width = Renderer.Width;
            _allMasterBarRenderers = new FastList<MasterBarsRenderers>();

            //
            // 1. Score Info
            y = LayoutAndRenderScoreInfo(x, y);

            //
            // 2. Chord Diagrms
            y = LayoutAndRenderChordDiagrams(y);

            //
            // 3. One result per StaveGroup
            y = LayoutAndRenderScore(x, y);

            Height = y + _pagePadding[3];
        }

        public override bool SupportsResize => true;

        public override void Resize()
        {
            var x = _pagePadding[0];
            var y = _pagePadding[1];
            Width = Renderer.Width;
            var oldHeight = Height;

            //
            // 1. Score Info
            y = LayoutAndRenderScoreInfo(x, y, oldHeight);

            //
            // 2. Chord Digrams
            y = LayoutAndRenderChordDiagrams(y, oldHeight);

            //
            // 2. One result per StaveGroup
            y = ResizeAndRenderScore(x, y, oldHeight);

            Height = y + _pagePadding[3];
        }

        private float LayoutAndRenderChordDiagrams(float y, float totalHeight = -1)
        {
            if (ChordDiagrams == null)
            {
                return y;
            }

            var res = Renderer.Settings.Display.RenderingResources;

            ChordDiagrams.Width = Width;
            ChordDiagrams.DoLayout();

            var canvas = Renderer.Canvas;
            canvas.BeginRender(Width, ChordDiagrams.Height);
            canvas.Color = res.ScoreInfoColor;
            canvas.TextAlign = TextAlign.Center;

            ChordDiagrams.Paint(0, 0, canvas);

            var result = canvas.EndRender();
            y += ChordDiagrams.Height;
            Renderer.OnPartialRenderFinished(new RenderFinishedEventArgs
            {
                Width = Width,
                Height = ChordDiagrams.Height,
                RenderResult = result,
                TotalWidth = Width,
                TotalHeight = totalHeight < 0 ? y : totalHeight,
                FirstMasterBarIndex = -1,
                LastMasterBarIndex = -1
            });

            return y;
        }

        private float LayoutAndRenderScoreInfo(float x, float y, float totalHeight = -1)
        {
            Logger.Debug(Name, "Layouting score info");

            var scale = Scale;
            var res = Renderer.Settings.Display.RenderingResources;

            var centeredGlyphs = new[]
            {
                HeaderFooterElements.Title, HeaderFooterElements.SubTitle, HeaderFooterElements.Artist,
                HeaderFooterElements.Album, HeaderFooterElements.WordsAndMusic
            };

            for (var i = 0; i < centeredGlyphs.Length; i++)
            {
                if (ScoreInfoGlyphs.ContainsKey(centeredGlyphs[i]))
                {
                    var glyph = ScoreInfoGlyphs[centeredGlyphs[i]];
                    glyph.X = Width / 2f;
                    glyph.Y = y;
                    glyph.TextAlign = TextAlign.Center;
                    y += glyph.Font.Size;
                }
            }

            var musicOrWords = false;
            float musicOrWordsHeight = 0;
            if (ScoreInfoGlyphs.ContainsKey(HeaderFooterElements.Music))
            {
                var glyph = ScoreInfoGlyphs[HeaderFooterElements.Music];
                glyph.X = Width - _pagePadding[2];
                glyph.Y = y;
                glyph.TextAlign = TextAlign.Right;
                musicOrWords = true;
                musicOrWordsHeight = glyph.Font.Size;
            }

            if (ScoreInfoGlyphs.ContainsKey(HeaderFooterElements.Words))
            {
                var glyph = ScoreInfoGlyphs[HeaderFooterElements.Words];
                glyph.X = x;
                glyph.Y = y;
                glyph.TextAlign = TextAlign.Left;
                musicOrWords = true;
                musicOrWordsHeight = glyph.Font.Size;
            }

            if (musicOrWords)
            {
                y += musicOrWordsHeight;
            }

            if (TuningGlyph != null)
            {
                y += 20 * scale;
                TuningGlyph.X = x;
                TuningGlyph.Y = y;
                y += TuningGlyph.Height;
            }

            y += 20 * scale;

            var canvas = Renderer.Canvas;
            canvas.BeginRender(Width, y);
            canvas.Color = res.ScoreInfoColor;
            canvas.TextAlign = TextAlign.Center;
            foreach (var key in ScoreInfoGlyphs)
            {
                ScoreInfoGlyphs[key].Paint(0, 0, canvas);
            }

            if (TuningGlyph != null)
            {
                TuningGlyph.Paint(0, 0, canvas);
            }

            var result = canvas.EndRender();
            Renderer.OnPartialRenderFinished(new RenderFinishedEventArgs
            {
                Width = Width,
                Height = y,
                RenderResult = result,
                TotalWidth = Width,
                TotalHeight = totalHeight < 0 ? y : totalHeight,
                FirstMasterBarIndex = -1,
                LastMasterBarIndex = -1
            });

            return y;
        }

        private float ResizeAndRenderScore(float x, float y, float oldHeight)
        {
            var canvas = Renderer.Canvas;

            // if we have a fixed number of bars per row, we only need to refit them.
            if (Renderer.Settings.Display.BarsPerRow != -1)
            {
                for (var i = 0; i < _groups.Count; i++)
                {
                    var group = _groups[i];
                    FitGroup(group);
                    group.FinalizeGroup();

                    y += PaintGroup(group, oldHeight, canvas);
                }
            }
            // if the bars per row are flexible, we need to recreate the stave groups
            // by readding the existing groups
            else
            {
                _groups = new FastList<StaveGroup>();

                var currentIndex = 0;
                var maxWidth = MaxWidth;

                var group = CreateEmptyStaveGroup();
                group.Index = _groups.Count;
                group.X = x;
                group.Y = y;

                while (currentIndex < _allMasterBarRenderers.Count)
                {
                    // if the current renderer still has space in the current group add it
                    // also force adding in case the group is empty
                    var renderers = _allMasterBarRenderers[currentIndex];
                    if (group.Width + renderers.Width <= maxWidth || group.MasterBarsRenderers.Count == 0)
                    {
                        group.AddMasterBarRenderers(Renderer.Tracks, renderers);
                        // move to next group
                        currentIndex++;
                    }
                    else
                    {
                        // if we cannot wrap on the current bar, we remove the last bar
                        // (this might even remove multiple ones until we reach a bar that can wrap);
                        while (renderers != null && !renderers.CanWrap && group.MasterBarsRenderers.Count > 1)
                        {
                            renderers = group.RevertLastBar();
                            currentIndex--;
                        }

                        // in case we do not have space, we create a new group
                        group.IsFull = true;
                        group.IsLast = LastBarIndex == group.LastBarIndex;
                        _groups.Add(group);
                        FitGroup(group);
                        group.FinalizeGroup();
                        y += PaintGroup(group, oldHeight, canvas);

                        // note: we do not increase currentIndex here to have it added to the next group
                        group = CreateEmptyStaveGroup();
                        group.Index = _groups.Count;
                        group.X = x;
                        group.Y = y;
                    }
                }

                group.IsLast = LastBarIndex == group.LastBarIndex;

                // don't forget to finish the last group
                FitGroup(group);
                group.FinalizeGroup();
                y += PaintGroup(group, oldHeight, canvas);
            }

            return y;
        }

        private float LayoutAndRenderScore(float x, float y)
        {
            var canvas = Renderer.Canvas;

            var startIndex = FirstBarIndex;
            var currentBarIndex = startIndex;

            var endBarIndex = LastBarIndex;

            _groups = new FastList<StaveGroup>();
            while (currentBarIndex <= endBarIndex)
            {
                // create group and align set proper coordinates
                var group = CreateStaveGroup(currentBarIndex, endBarIndex);
                _groups.Add(group);
                group.X = x;
                group.Y = y;

                currentBarIndex = group.LastBarIndex + 1;

                // finalize group (sizing etc).
                FitGroup(group);
                group.FinalizeGroup();

                Logger.Info(Name, "Rendering partial from bar " + group.FirstBarIndex + " to " + group.LastBarIndex);
                y += PaintGroup(group, y, canvas);
            }

            return y;
        }

        private float PaintGroup(StaveGroup group, float totalHeight, ICanvas canvas)
        {
            // paint into canvas
            var height = group.Height + GroupSpacing * Scale;
            canvas.BeginRender(Width, height);
            Renderer.Canvas.Color = Renderer.Settings.Display.RenderingResources.MainGlyphColor;
            Renderer.Canvas.TextAlign = TextAlign.Left;
            // NOTE: we use this negation trick to make the group paint itself to 0/0 coordinates
            // since we use partial drawing
            group.Paint(0, -group.Y, canvas);

            // calculate coordinates for next group
            totalHeight += height;

            var result = canvas.EndRender();
            var args = new RenderFinishedEventArgs();
            args.TotalWidth = Width;
            args.TotalHeight = totalHeight;
            args.Width = Width;
            args.Height = height;
            args.RenderResult = result;
            args.FirstMasterBarIndex = group.FirstBarIndex;
            args.LastMasterBarIndex = group.LastBarIndex;

            Renderer.OnPartialRenderFinished(args);

            return height;
        }

        /// <summary>
        /// Realignes the bars in this line according to the available space
        /// </summary>
        private void FitGroup(StaveGroup group)
        {
            if (group.IsFull || group.Width > MaxWidth)
            {
                group.ScaleToWidth(MaxWidth);
            }

            Width = Math.Max(Width, group.Width);
        }

        private StaveGroup CreateStaveGroup(int currentBarIndex, int endIndex)
        {
            var group = CreateEmptyStaveGroup();
            group.Index = _groups.Count;

            var barsPerRow = Renderer.Settings.Display.BarsPerRow;

            var maxWidth = MaxWidth;
            var end = endIndex + 1;
            for (var i = currentBarIndex; i < end; i++)
            {
                if (_barsFromPreviousGroup.Count > 0)
                {
                    foreach (var renderer in _barsFromPreviousGroup)
                    {
                        group.AddMasterBarRenderers(Renderer.Tracks, renderer);
                        i = renderer.MasterBar.Index;
                    }
                }
                else
                {
                    var renderers = group.AddBars(Renderer.Tracks, i);
                    _allMasterBarRenderers.Add(renderers);
                }

                _barsFromPreviousGroup = new FastList<MasterBarsRenderers>();

                var groupIsFull = false;

                // can bar placed in this line?
                if (barsPerRow == -1 && group.Width >= maxWidth && group.MasterBarsRenderers.Count != 0)
                {
                    groupIsFull = true;
                }
                else if (group.MasterBarsRenderers.Count == barsPerRow + 1)
                {
                    groupIsFull = true;
                }

                if (groupIsFull)
                {
                    var reverted = group.RevertLastBar();
                    if (reverted != null)
                    {
                        _barsFromPreviousGroup.Add(reverted);

                        while (reverted != null && !reverted.CanWrap && group.MasterBarsRenderers.Count > 1)
                        {
                            reverted = group.RevertLastBar();
                            _barsFromPreviousGroup.Add(reverted);
                        }
                    }

                    group.IsFull = true;
                    group.IsLast = false;

                    _barsFromPreviousGroup.Reverse();
                    return group;
                }

                group.X = 0;
            }

            group.IsLast = endIndex == group.LastBarIndex;
            return group;
        }

        private float MaxWidth => Renderer.Width - _pagePadding[0] - _pagePadding[2];
    }
}

File: src/rendering/layout/PageViewLayout.ts
import { type ICanvas, TextAlign } from '@src/platform/ICanvas';
import type { TextGlyph } from '@src/rendering/glyphs/TextGlyph';
import { InternalSystemsLayoutMode, ScoreLayout } from '@src/rendering/layout/ScoreLayout';
import { RenderFinishedEventArgs } from '@src/rendering/RenderFinishedEventArgs';
import type { MasterBarsRenderers } from '@src/rendering/staves/MasterBarsRenderers';
import type { StaffSystem } from '@src/rendering/staves/StaffSystem';
import type { RenderingResources } from '@src/RenderingResources';
import { Logger } from '@src/Logger';
import { SystemsLayoutMode } from '@src/DisplaySettings';
import { ScoreSubElement } from '@src/model/Score';

/**
 * This layout arranges the bars into a fixed width and dynamic height region.
 */
export class PageViewLayout extends ScoreLayout {
    private _systems: StaffSystem[] = [];
    private _allMasterBarRenderers: MasterBarsRenderers[] = [];
    private _barsFromPreviousSystem: MasterBarsRenderers[] = [];

    public get name(): string {
        return 'PageView';
    }

    protected doLayoutAndRender(): void {
        switch (this.renderer.settings.display.systemsLayoutMode) {
            case SystemsLayoutMode.Automatic:
                this.systemsLayoutMode = InternalSystemsLayoutMode.Automatic;
                break;
            case SystemsLayoutMode.UseModelLayout:
                this.systemsLayoutMode = InternalSystemsLayoutMode.FromModelWithScale;
                break;
        }

        let y: number = this.pagePadding![1];
        this.width = this.renderer.width;
        this._allMasterBarRenderers = [];
        //
        // 1. Score Info
        y = this.layoutAndRenderScoreInfo(y, -1);
        //
        // 2. Tunings
        y = this.layoutAndRenderTunings(y, -1);
        //
        // 3. Chord Diagrms
        y = this.layoutAndRenderChordDiagrams(y, -1);
        //
        // 4. One result per StaffSystem
        y = this.layoutAndRenderScore(y);

        y = this.layoutAndRenderBottomScoreInfo(y);

        y = this.layoutAndRenderAnnotation(y);

        this.height = (y + this.pagePadding![3]) * this.renderer.settings.display.scale;
    }

    public get supportsResize(): boolean {
        return true;
    }

    public get firstBarX(): number {
        let x = this.pagePadding![0];
        if (this._systems.length > 0) {
            x += this._systems[0].accoladeWidth;
        }
        return x;
    }

    public doResize(): void {
        let y: number = this.pagePadding![1];
        this.width = this.renderer.width;
        const oldHeight: number = this.height;
        //
        // 1. Score Info
        y = this.layoutAndRenderScoreInfo(y, oldHeight);
        //
        // 2. Tunings
        y = this.layoutAndRenderTunings(y, oldHeight);
        //
        // 3. Chord Digrams
        y = this.layoutAndRenderChordDiagrams(y, oldHeight);
        //
        // 4. One result per StaffSystem
        y = this.resizeAndRenderScore(y, oldHeight);

        y = this.layoutAndRenderBottomScoreInfo(y);

        y = this.layoutAndRenderAnnotation(y);

        this.height = (y + this.pagePadding![3]) * this.renderer.settings.display.scale;
    }

    private layoutAndRenderTunings(y: number, totalHeight: number = -1): number {
        if (!this.tuningGlyph) {
            return y;
        }

        const res: RenderingResources = this.renderer.settings.display.resources;
        this.tuningGlyph.x = this.pagePadding![0];
        this.tuningGlyph.width = this.scaledWidth;
        this.tuningGlyph.doLayout();

        const tuningHeight = Math.round(this.tuningGlyph.height);

        const e = new RenderFinishedEventArgs();
        e.x = 0;
        e.y = y;
        e.width = this.scaledWidth;
        e.height = tuningHeight;
        e.totalWidth = this.scaledWidth;
        e.totalHeight = totalHeight < 0 ? y + e.height : totalHeight;

        this.registerPartial(e, (canvas: ICanvas) => {
            canvas.color = res.scoreInfoColor;
            canvas.textAlign = TextAlign.Center;
            this.tuningGlyph!.paint(0, 0, canvas);
        });

        return y + tuningHeight;
    }

    private layoutAndRenderChordDiagrams(y: number, totalHeight: number = -1): number {
        if (!this.chordDiagrams) {
            return y;
        }
        const res: RenderingResources = this.renderer.settings.display.resources;
        this.chordDiagrams.width = this.scaledWidth;
        this.chordDiagrams.doLayout();

        const diagramHeight = Math.round(this.chordDiagrams.height);

        const e = new RenderFinishedEventArgs();
        e.x = 0;
        e.y = y;
        e.width = this.scaledWidth;
        e.height = diagramHeight;
        e.totalWidth = this.scaledWidth;
        e.totalHeight = totalHeight < 0 ? y + diagramHeight : totalHeight;

        this.registerPartial(e, (canvas: ICanvas) => {
            canvas.color = res.scoreInfoColor;
            canvas.textAlign = TextAlign.Center;
            this.chordDiagrams!.paint(0, 0, canvas);
        });

        return y + diagramHeight;
    }

    private layoutAndRenderScoreInfo(y: number, totalHeight: number = -1): number {
        Logger.debug(this.name, 'Layouting score info');

        const e = new RenderFinishedEventArgs();
        e.x = 0;
        e.y = y;

        let infoHeight = 0;

        const res: RenderingResources = this.renderer.settings.display.resources;

        const scoreInfoGlyphs: TextGlyph[] = [];

        for (const [scoreElement, _notationElement] of ScoreLayout.HeaderElements.value) {
            if (this.headerGlyphs.has(scoreElement)) {
                const glyph: TextGlyph = this.headerGlyphs.get(scoreElement)!;
                glyph.y = infoHeight;
                this.alignScoreInfoGlyph(glyph);

                let lineHeight = glyph.font.size;

                // words and music on same line if not aligned on same side
                if (scoreElement === ScoreSubElement.Words) {
                    if (this.headerGlyphs.has(ScoreSubElement.Music)) {
                        const musicGlyph = this.headerGlyphs.get(ScoreSubElement.Music)!;
                        if (musicGlyph.textAlign !== glyph.textAlign) {
                            lineHeight = 0;
                        }
                    }
                }

                infoHeight += lineHeight;

                scoreInfoGlyphs.push(glyph);
            }
        }

        if (scoreInfoGlyphs.length > 0) {
            infoHeight = Math.floor(infoHeight + 17);
            e.width = this.scaledWidth;
            e.height = infoHeight;
            e.totalWidth = this.scaledWidth;
            e.totalHeight = totalHeight < 0 ? y + e.height : totalHeight;
            this.registerPartial(e, (canvas: ICanvas) => {
                canvas.color = res.scoreInfoColor;
                canvas.textAlign = TextAlign.Center;
                for (const g of scoreInfoGlyphs) {
                    g.paint(0, 0, canvas);
                }
            });
        }

        return y + infoHeight;
    }

    private resizeAndRenderScore(y: number, oldHeight: number): number {
        // if we have a fixed number of bars per row, we only need to refit them.
        const barsPerRowActive =
            this.renderer.settings.display.barsPerRow > 0 ||
            this.systemsLayoutMode === InternalSystemsLayoutMode.FromModelWithScale;

        if (barsPerRowActive) {
            for (let i: number = 0; i < this._systems.length; i++) {
                const system: StaffSystem = this._systems[i];
                this.fitSystem(system);
                y += this.paintSystem(system, oldHeight);
            }
        } else {
            this._systems = [];
            let currentIndex: number = 0;
            const maxWidth: number = this.maxWidth;
            let system: StaffSystem = this.createEmptyStaffSystem();
            system.index = this._systems.length;
            system.x = this.pagePadding![0];
            system.y = y;
            while (currentIndex < this._allMasterBarRenderers.length) {
                // if the current renderer still has space in the current system add it
                // also force adding in case the system is empty
                let renderers: MasterBarsRenderers | null = this._allMasterBarRenderers[currentIndex];
                if (system.width + renderers!.width <= maxWidth || system.masterBarsRenderers.length === 0) {
                    system.addMasterBarRenderers(this.renderer.tracks!, renderers!);
                    // move to next system
                    currentIndex++;
                } else {
                    // if we cannot wrap on the current bar, we remove the last bar
                    // (this might even remove multiple ones until we reach a bar that can wrap);
                    while (renderers && !renderers.canWrap && system.masterBarsRenderers.length > 1) {
                        renderers = system.revertLastBar();
                        currentIndex--;
                    }
                    // in case we do not have space, we create a new system
                    system.isFull = true;
                    system.isLast = this.lastBarIndex === system.lastBarIndex;
                    this._systems.push(system);
                    this.fitSystem(system);
                    y += this.paintSystem(system, oldHeight);
                    // note: we do not increase currentIndex here to have it added to the next system
                    system = this.createEmptyStaffSystem();
                    system.index = this._systems.length;
                    system.x = this.pagePadding![0];
                    system.y = y;
                }
            }
            system.isLast = this.lastBarIndex === system.lastBarIndex;
            // don't forget to finish the last system
            this.fitSystem(system);
            y += this.paintSystem(system, oldHeight);
        }
        return y;
    }

    private layoutAndRenderScore(y: number): number {
        const startIndex: number = this.firstBarIndex;
        let currentBarIndex: number = startIndex;
        const endBarIndex: number = this.lastBarIndex;

        this._systems = [];
        while (currentBarIndex <= endBarIndex) {
            // create system and align set proper coordinates
            const system: StaffSystem = this.createStaffSystem(currentBarIndex, endBarIndex);
            this._systems.push(system);
            system.x = this.pagePadding![0];
            system.y = y;
            currentBarIndex = system.lastBarIndex + 1;
            // finalize system (sizing etc).
            this.fitSystem(system);
            Logger.debug(
                this.name,
                `Rendering partial from bar ${system.firstBarIndex} to ${system.lastBarIndex}`,
                null
            );
            y += this.paintSystem(system, y);
        }
        return y;
    }

    private paintSystem(system: StaffSystem, totalHeight: number): number {
        // paint into canvas
        const height: number = Math.floor(system.height);

        const args: RenderFinishedEventArgs = new RenderFinishedEventArgs();
        args.x = 0;
        args.y = system.y;
        args.totalWidth = this.scaledWidth;
        args.totalHeight = totalHeight;
        args.width = this.scaledWidth;
        args.height = height;
        args.firstMasterBarIndex = system.firstBarIndex;
        args.lastMasterBarIndex = system.lastBarIndex;

        system.buildBoundingsLookup(0, 0);
        this.registerPartial(args, canvas => {
            this.renderer.canvas!.color = this.renderer.settings.display.resources.mainGlyphColor;
            this.renderer.canvas!.textAlign = TextAlign.Left;
            // NOTE: we use this negation trick to make the system paint itself to 0/0 coordinates
            // since we use partial drawing
            system.paint(0, -(args.y / this.renderer.settings.display.scale), canvas);
        });

        // calculate coordinates for next system
        return height;
    }

    /**
     * Realignes the bars in this line according to the available space
     */
    private fitSystem(system: StaffSystem): void {
        if (system.isFull || system.width > this.maxWidth || this.renderer.settings.display.justifyLastSystem) {
            system.scaleToWidth(this.maxWidth);
        } else {
            system.scaleToWidth(system.width);
        }
        system.finalizeSystem();
    }

    private getBarsPerSystem(rowIndex: number) {
        let barsPerRow: number = this.renderer.settings.display.barsPerRow;

        if (this.systemsLayoutMode === InternalSystemsLayoutMode.FromModelWithScale) {
            let defaultSystemsLayout: number;
            let systemsLayout: number[];
            if (this.renderer.tracks!.length > 1) {
                // multi track applies
                defaultSystemsLayout = this.renderer.score!.defaultSystemsLayout;
                systemsLayout = this.renderer.score!.systemsLayout;
            } else {
                defaultSystemsLayout = this.renderer.tracks![0].defaultSystemsLayout;
                systemsLayout = this.renderer.tracks![0].systemsLayout;
            }

            barsPerRow = rowIndex < systemsLayout.length ? systemsLayout[rowIndex] : defaultSystemsLayout;
        }

        return barsPerRow;
    }

    private createStaffSystem(currentBarIndex: number, endIndex: number): StaffSystem {
        const system: StaffSystem = this.createEmptyStaffSystem();
        system.index = this._systems.length;
        const barsPerRow: number = this.getBarsPerSystem(system.index);
        const maxWidth: number = this.maxWidth;
        const end: number = endIndex + 1;

        let barIndex = currentBarIndex;
        while (barIndex < end) {
            if (this._barsFromPreviousSystem.length > 0) {
                for (const renderer of this._barsFromPreviousSystem) {
                    system.addMasterBarRenderers(this.renderer.tracks!, renderer);
                    barIndex = renderer.lastMasterBarIndex;
                }
            } else {
                const multiBarRestInfo = this.multiBarRestInfo;
                const additionalMultiBarsRestBarIndices: number[] | null =
                    multiBarRestInfo !== null && multiBarRestInfo.has(barIndex)
                        ? multiBarRestInfo.get(barIndex)!
                        : null;

                const renderers = system.addBars(this.renderer.tracks!, barIndex, additionalMultiBarsRestBarIndices);
                this._allMasterBarRenderers.push(renderers);
                barIndex = renderers.lastMasterBarIndex;
            }
            this._barsFromPreviousSystem = [];
            let systemIsFull: boolean = false;
            // can bar placed in this line?
            if (barsPerRow === -1 && system.width >= maxWidth && system.masterBarsRenderers.length !== 0) {
                systemIsFull = true;
            } else if (system.masterBarsRenderers.length === barsPerRow + 1) {
                systemIsFull = true;
            }
            if (systemIsFull) {
                let reverted = system.revertLastBar();
                if (reverted) {
                    this._barsFromPreviousSystem.push(reverted);
                    while (reverted && !reverted.canWrap && system.masterBarsRenderers.length > 1) {
                        reverted = system.revertLastBar();
                        if (reverted) {
                            this._barsFromPreviousSystem.push(reverted);
                        }
                    }
                }
                system.isFull = true;
                system.isLast = false;
                this._barsFromPreviousSystem.reverse();
                return system;
            }
            // do we need a line break after this bar
            let anyTrackNeedsLineBreak = false;
            let allTracksNeedLineBreak = true;
            for (const track of this.renderer.tracks!) {
                if (track.lineBreaks && track.lineBreaks!.has(barIndex + 1)) {
                    anyTrackNeedsLineBreak = true;
                } else {
                    allTracksNeedLineBreak = false;
                }
            }

            if (anyTrackNeedsLineBreak && allTracksNeedLineBreak) {
                system.isFull = true;
                system.isLast = false;
                return system;
            }
            system.x = 0;
            barIndex++;
        }
        system.isLast = endIndex === system.lastBarIndex;
        return system;
    }

    private get maxWidth(): number {
        return this.scaledWidth - this.pagePadding![0] - this.pagePadding![2];
    }
}
